virtual_rainforest/core/base_model.py

```diff
@@ -119,34 +163,58 @@ def cleanup(self) -> None:
 
     @classmethod
     @abstractmethod
-    def from_config(cls, config: dict[str, Any]) -> Any:
+    def from_config(cls, data: Data, config: dict[str, Any]) -> BaseModel:
         """Factory function to unpack config and initialise a model instance."""
 
     @classmethod
     def __init_subclass__(cls) -> None:
-        """Method that adds new model classes to the model registry.
+        """Initialise subclasses deriving from BaseModel.
+
+        This method runs when a new BaseModel subclass is imported. It adds the new
+        subclasses to the model registry and validates the values of the class
+        properties.
 
         Raises:
-            ValueError: If model_name attribute isn't defined
+            ValueError: If the model_name or required_init_vars properties are not
+                defined
             TypeError: If model_name is not a string
         """
 
-        # Check that model_name exists and is a string
-        if not hasattr(cls, "model_name"):
-            excep = ValueError("Models must have a model_name attribute!")
+        excep: Exception
+
+        # Check that model_name exists and is a string - if it is not implemented in the
+        # subclass then the object will be of type property
+        if isinstance(cls.model_name, property):
+            excep = NotImplementedError(
+                f"Property model_name is not implemented in {cls.__name__}"
+            )
```

Comment:
Given that `model_name` is an `abstractmethod`, shouldn't be just impossible to create a subclass without overwriting it? That's the whole point of `abstractmethods`, right?

Reply:
I got hung up on this too! The distinction is (hat tip, @jacobcook1995) that the automatic testing for missing `abstractmethod` objects only happens when you try and create an instance of the class. When `__init_subclass__` runs, it is just dealing with the `cls` object. In this function, we want to check the name is valid, but do need to gatekeep first for undefined `model_name`. 

Arguably, we could leave out the `required_init_vars` - we don't currently test the contents - and then it would get flagged at instantiation if missing. I think testing the contents is probably valuable though.

Reply:
Interesting. Indeed, I didn't know abstract classes worked that way. Yes, I agree it is useful to check during the definition of the subclasses.

Reply:
Is this kind of structure using `assert` ok for checking the contents of `required_init_vars`. I've not really used `assert` like this before but it seems a compact way to raise exceptions without having to nest a bunch of `if` statements to catch specific failure modes:

```python
        # Check the structure
        try:
            assert isinstance(cls.required_init_vars, tuple)
            for vname, axes in cls.required_init_vars:
                assert isinstance(vname, str)
                assert isinstance(axes, tuple)
                assert all([isinstance(a, str) for a in axes])
        except (AssertionError, ValueError):
            raise ValueError(
                f"Property required_init_vars has the wrong structure in {cls.__name__}"
            )
```

Reply:
There is nothing conceptually wrong with using `assert` I think, but `assert` checks are disabled when running python with the `-O` flag. From `python -h`:

```bash
-O     : remove assert and __debug__-dependent statements;
```

I'm not sure if you intend to run the code in optimised mode, but just in case I would not use `assert` to control the flow of the code.

Reply:
Ah. Damn. Yes - we don't want error checking to disappear in optimised mode. More long winded approach it is then.